
# Chloroplast annotation

List of tools that can be used


Prokka => No intron allowed

Geseq which is web based tool and easy to use. It works well with both mitochondria and chloroplast.

AGORA - organelle genome annotation
Jung J, Kim JI, Jeong Y-S and Yi, G (2018) AGORA: organellar genome annotation from the amino acid and nucleotide references. Bioinformatics 34: 2661-2663

CpGAVAS - chloroplast genome annotation
Liu C, Shi L, Zhu Y, Chen H, Zhang J, Lin X and Guan X (2012) CpGAVAS, an integrated web server for the annotation, visualization, analysis, and GenBank submission of completely sequenced chloroplast genome sequences. BMC Genomics 13: 715

DOGMA - plant chloroplast and animal mitochondrial genome annotation
Wyman SK, Jansen RK and Boore JL (2004) Automatic annotation of organellar genomes with DOGMA. Bioinformatics 20: 3252-3255

MFannot - organelle genome annotation
Developed by the labs of B. Franz Lang and Gertraud Burger (unpublished)
output not easy to deal with. Here the tool we use to convert in gff mfannot2gff.pl. But introns type II are not in that output.

ORG.Annotate - chloroplast genome annotation
Developed by Eric Coissac (unpublished)

Plann - chloroplast genome annotation
Huang DI and Cronk QCB (2015) Plann: a command-line application for annotating plastome sequences (2015). Applications in Plant Sciences 3: 1500026

Verdant - chloroplast genome annotation and phylogenetic analysis
McKain MR, Hartsock RH, Wohl MM and Kellogg EA (2017) Verdant: automated annotation, alignment and phylogenetic analysis of whole chloroplast genomes. Bioinformatics 33: 130-132

(Nice resource where we catch most of those tools https://chlorobox.mpimp-golm.mpg.de/Alternative-Tools.html)
