| year	| Gene finder Name | Publication | Type	| Method | Organism | Nb citation (pubmed 2016) | Comments | Output Format |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
1991 | GRAIL |	| Ab initio | | | | No longer supported |
1992 | GeneID |	Gugio et al. | Ab initio | WAM, HMM, PD, AD | | | |			
1993 | GeneParser | |	Ab initio	| | | | |		
1993 | GRAIL-II | Milanesi et al. | | | | | |
1994 | Fgeneh	| | Ab initio | HMM, DP, LDA | Eukaryota | | Finds single exon only |
1994 | EcoParse	| | Ab initio | HMM | Prokaryote | 393 | |
1994 | GeneLang | Dong and Searls | Ab initio | Linguistic method HMM, PD, WAM | Eukaryote | | | |
1995 | FGENES| Solovyev et al. | | | | | | |
1996 | Genie	| Kulp et al. | Hybrid | | | | |			
1996 | PROCRUSTES | | Evidence based | | | | |			
1997 | Fgenes	| | Hybrid | | | | | No download version |
1997 | GeneFinder | | Ab initio | | | | Unpublished work |
1997 | GenScan	| Burge and Karlin | Ab initio | | | | | |
1997 | MZEF	| Zhang | | | | | | |
1997 | HMMGene | Kough | Ab initio | | | | No download version | 
1997 | GeneWise	| | Evidence based | | | | | 		
1997 | AAT (Analysis and Annotation Tool) | Huang et al. | Evidence based | | | | Include two paris of programs DPS/NAP and DDS/GAP | 
1998 | GeneMark.hmm | Lukashin, A. V & Borodovsky, M. GeneMark. hmm: new solutions for gene finding. Nucleic Acids Res. 26, 1107–1115 (1998). | Ab initio | HMM | Prokaryote | 1334 | Self training |
1999 | Glimmer | Delcher et al. | | | Prokartyote | | |
1999 | GlimmerM | Salzberg,S.L., Pertea,M., Delcher,A.L., Gardner,M.J. and Tettelin,H. (1999) Interpolated Markov models for eukaryotic gene finding. Genomics, 59, 24–31. | | | Small eukaryote | | developed to find genes in the malaria parasite Plasmodium falciparum. |
1999 | Veil (the Viterbi Exon-Intron Locator) | Finding Genes in Human DNA with a Hidden Markov Model. J. Henderson, S.L. Salzberg, and K. Fasman. This describes the VEIL system for finding genes. Journal of Computational Biology 4:2 (1997), 127-141. | HMM | | | Eukaryote | | 
2000 | GenomeScan | | Comparative | | | | |	
2001 | Eugene |	| Hybrid | Semi-Markov Conditional Random Fields / IMM, DP | Plant | | Can be seen as a combiner because collect information about splice sites and ATG has to be done outside the program. |
2001 | Twinscan | | Comparative	| | | | Two genomes |
2001 | GeneHacker Plus | | Ab initio | HMM | Prokaryote | 50 | |	
2001 | GeneMarkS | | Ab initio | HMM | Prokaryote | 742 |	Self training |
2002 | GAZE	| Howe, K. L. et al. GAZE : A Generic Framework for the Integration of Gene-Prediction Data by Dynamic Programming. 1418–1427 (2002). doi:10.1101/gr.149502 | Comparative / combiner | | | | |		
2003 | SLAM	| | Comparative |	GHMM | | 187 | |	
2003 | EasyGene | | Ab initio |	HMM, H | Prokaryote |	153	| |
2003 | ETOPE | | Comparative | based on the ratio of non-synonymous to synonymous substitution rates between sequences from different genomes	| Eucaryote |	20	| Based on Genscan output |
2004 | Ensembl | | Evidence based	| | |	| Pipeline |
2004 | GeneZilla/TIGRSCAN | |	Ab initio | | | |	No longer supported |
2004 | GlimmerHMM	| Majoros, W.H., Pertea, M.,and Salzberg, S.L. TigrScan and GlimmerHMM: two open-source ab initio eukaryotic gene-finders Bioinformatics 2004 2878-2879. | Ab initio	| GHMM | eukaryote | | |		
2004 | TigrScan | Majoros, W.H., Pertea, M.,and Salzberg, S.L. TigrScan and GlimmerHMM: two open-source ab initio eukaryotic gene-finders Bioinformatics 2004 2878-2879. | Ab initio	| GHMM | eukaryote | | |		
2004 | SNAP |	| Ab initio | | | |			
2005 | JIGSAW	| |  Combiner | | | 137 | select the prediction whose structure best represents the consensus |
2005 | GeneMark-ES | Lomsadze, A. Gene identification in novel eukaryotic genomes by self-training algorithm. Nucleic Acids Res. 33, 6494–6506 (2005) ; Ter-Hovhannisyan, V., Lomsadze, A., Chernoff, Y. O. & Borodovsky, M. Gene prediction in novel fungal genomes using an ab initio algorithm with unsupervised training. Genome Res. 18, 1979–1990 (2008). | Ab initio |  | Eucaryote | 243 / 200 | |
2006 | AUGUSTUS+ | | Hybrid | HMM | | |			
2006 | N-SCAN	| | Comparative |	| | |	Several genomes |
2006 | TWINSCAN_EST | | Comparative+Evidence | | | | |
2006 | N_Scan_EST | | Comparative+Evidence | | | | |		
2006 | Metagene	| | | | Metagenomic	| 294 | |	
2007 | Conrad | | Ab initio | | | | |				
2007 | Contrast | | Comparative | | | 90 | Can also incorporate information from EST alignment |
2007 | Genomix	| | | | |	1 | |	
2008 | Maker | Cantarel, B. L. et al. Maker. Genome Res. 18, 188–96 (2008). | Combiner | | | 306	| It uses proteins, transcripts ... Abinitio: Augustus, Fgnesh,Genemark,snap|
2008 | Evigan | Liu, Q., Mackey, A. J., Roos, D. S. & Pereira, F. C. N. Evigan: A hidden variable model for integrating gene evidence for eukaryotic gene prediction. Bioinformatics 24, 597–605 (2008). | Combiner |	Unsupervised learning method | | 52 | Choose the best possible set of exons and combine them in a gene model. Weight of different sources.|
2008 | Evidence Modeler | Haas, B. J. et al. Automated eukaryotic gene structure annotation using EVidenceModeler and the Program to Assemble Spliced Alignments. Genome Biol. 9, R7 (2008). | Combiner | | | | choose the best possible set of exons and combine them in a gene model weight of different sources. Evidence based chooser. | 
2009 | mGene | Schweikert, G. et al. mGene: accurate SVM-based gene finding with an application to nematode genomes. Genome Res. 19, 2133–43 (2009). | Ab initio | SVM	| | 66 | No longer supported |
2009 | Orphelia | |	Ab initio |	Neural network | Metagenomic | 78 | | |
2009 | DAWGPAWS | Estill, J. C. & Bennetzen, J. L. The DAWGPAWS pipeline for the annotation of genes and transposable elements in plant genomes. Plant Methods 5, 1–11 (2009). | | | Eukaryote / Plant  | | pipeline for the annotation of genes and transposable elements in plant genomes | |
2010 | MetaGeneMark | Zhu, W., Lomsadze, A. & Borodovsky, M. Ab initio gene identification in metagenomic sequences. Nucleic Acids Res. 38, 1–15 (2010). | Ab initio | HMM | Metagenome | 220 | Self training |
2011 | MAKER2 | Holt, C. & Yandell, M. MAKER2: an annotation pipeline and genome-database management tool for second-generation genome projects. BMC Bioinformatics 12, 491 (2011). | | | | 184 | It uses proteins, transcripts ... Abinitio: Augustus, Fgnesh,Genemark,snap |	
2012 | eCRAIG | | | | | 4 | |	
2014 | GeneMark-ET| Lomsadze, A., Burns, P. D. & Borodovsky, M. Integration of mapped RNA-Seq reads into automatic training of eukaryotic gene finding algorithm. 42, 1–8 (2014). | Ab initio | HMM | Eukaryote | 10 | Self training |
2015 | Ipred | Zickmann, F. & Renard, B. Y. IPred - integrating ab initio and evidence based gene predictions to improve prediction accuracy. BMC Genomics 16, 134 (2015). | Combiner evidence-based | | | | choose the best possible set of exons and combine them in a gene model. Evidence based chooser. Can also model gene form evidence only. |
2016 | BRAKER1 | Lange, S., Hof, K., Lomsadze, A., Borodovsky, M. & Stanke, M. BRAKER1 : Unsupervised RNA-Seq-Based Genome Annotation with GeneMark-ET and AUGUSTUS. 24, 2014 (2014). | Hybrid / Pipeline	| | | 1 | Pipeline: GeneMark-ET + Augustus |	
2016 | Companion | Steinbiss, S. et al. Companion: a web server for annotation and analysis of parasite genomes. Nucleic Acids Res. 44, W29–W34 (2016) | | | | | Pipeline for automatic eukaryotic parasite annotation |
2016 | Gmove | Dubarry, M. et al. Gmove a tool for eukaryotic gene predictions using various evidences. F1000Reserach 34, 2011 (2016). | | | | | Eukaryotic Gene Predictions using Various Evidences Dubarry |
2017 | GeMoMa | Keilwagen, J., Hartung, F., Paulini, M., Twardziok, S. O. & Grau, J. Combining RNA-seq data and homology-based gene prediction for plants , animals and fungi. (2017). | | | | | homology-based gene prediction program				
_____________________
__Legend:__  
Hybrid = ab initio and evidence based;  
Comparative = genome sequence comparison  
  
  
CHMM: class HMM  
DP: dynamic programming  
GHMM: generalized HMM  
HMM: hidden MM  
IMM: Interpolated MM   
MDD: maximal dependence decomposition  
ML: maximum likelihood   
MM: Markov Model  
NN: Neural Networks   
WAM: weight array matrix  
LDA: Linear Discriminant Analysis  
(baczkowski review)  

















