
# Mitochondrial annotation

List of tools that can be used

MitoFish => adapted for fish genome

Prokka => No intron allowed

MAKER (with specific codon table) => not really good but provide nice protein alignment useful for the manual curation

Geseq which is web based tool and easy to use. It works well with both mitochondria and chloroplast.

AGORA - organelle genome annotation
Jung J, Kim JI, Jeong Y-S and Yi, G (2018) AGORA: organellar genome annotation from the amino acid and nucleotide references. Bioinformatics 34: 2661-2663

MFannot - organelle genome annotation
Developed by the labs of B. Franz Lang and Gertraud Burger (unpublished)
output not easy to deal with. Here the tool we use to convert in gff mfannot2gff.pl. But introns type II are not in that output.

*MitoAnnotator*​ - fish mitochondrial genome annotation
Iwasaki W, Fukunaga T, Isagozawa R, Yamada K, Maeda, Y, Satoh TP, Sado T, Mabuchi K, Takeshima H, Miya M and Nishida M (2013) MitoFish and MitoAnnotator: a mitochondrial genome database of fish with an accurate and automatic annotation pipeline. Molecular Biology and Evolution 30: 2531-2540

MITOFY - plant mitochondrial genome annotation
Alverson AJ, Wei X, Rice DW, Stern DB, Barry K and Palmer JD (2010) Insights into the evolution of mitochondrial genome size from complete sequences of Citrullus lanatus and Cucurbita pepo (Cucurbitaceae). Molecular Biology and Evolution 27: 1436-1448

MITOS - metazoan mitochondrial genome annotation
Bernt M, Donath A, Jühling F, Externbrink F, Florentz C, Fritzsch G, Pütz J, Middendorf M and Stadler PF (2013) MITOS: improved de novo metazoan mitochondrial genome annotation. Molecular Phylogenetics and Evolution 69: 313-319

MOSAS - insect mitochondrial genome annotation
Sheffield NC, Hiatt KD, Valentine MC, Song H and Whiting MF (2010) Mitochondrial genomics in Orthoptera using MOSAS. Mitochondrial DNA 21: 87-104

(Nice resource where we catch most of those tools https://chlorobox.mpimp-golm.mpg.de/Alternative-Tools.html)
